lib/recursive-methods.rb

```diff
@@ -1,49 +1,117 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0 || n == 1
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  end
+  
+  until n == 1 do
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s
+  end
+  
+  reversed_str = reverse(s[1..-1])
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.